lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py

```diff
@@ -0,0 +1,178 @@
+        )
+
+    def _on_provider_endpoints_relation_created(self, event: RelationCreatedEvent) -> None:
+        self.on.ready.emit()
+
+    def send_endpoints_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "consent_url": f"{endpoint}/consent",
+                    "error_url": f"{endpoint}/error",
+                    "index_url": f"{endpoint}/index",
+                    "login_url": f"{endpoint}/login",
+                    "oidc_error_url": f"{endpoint}/oidc_error",
+                    "registration_url": f"{endpoint}/registration",
+                    "default_url": endpoint,
+                }
+            )
+
+
+class LoginUIEndpointsRelationError(Exception):
+    """Base class for the relation exceptions."""
+
+    pass
+
+
+class LoginUIEndpointsRelationMissingError(LoginUIEndpointsRelationError):
+    """Raised when the relation is missing."""
+
+    def __init__(self) -> None:
+        self.message = "Missing ui-endpoint-info relation with Identity Platform Login UI"
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationDataMissingError(LoginUIEndpointsRelationError):
+    """Raised when information is missing from the relation."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRelationUnavailableError(LoginUIEndpointsRelationError):
+    """Raised when Login UI cannot be accessed."""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class LoginUIEndpointsRequirer(Object):
+    """Requirer side of the ui-endpoint-info relation."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self._relation_name = relation_name
+
+    def get_login_ui_endpoints(self) -> Optional[Dict]:
+        """Get the Identity Platform Login UI endpoints."""
+        if not self.model.unit.is_leader():
+            return None
+        endpoints = self.model.relations[self._relation_name]
+        if len(endpoints) == 0:
+            raise LoginUIEndpointsRelationMissingError()
+
+        data = endpoints[0].data[endpoints[0].app]
+                
+        if any(not data.get(k := key) for key in RELATION_KEYS):
+            raise LoginUIEndpointsRelationDataMissingError(
+                f"Missing endpoint {k} in ui-endpoint-info relation data"
+            )
+
+        if data["default_url"] == "":
+            raise LoginUIEndpointsRelationUnavailableError(
+                "Endpoints in ui-endpoint-info are unavailable"
+            )
```

Comment:
when the proxy for login_ui goes down, the default_browser gets updated to empty string. so here we're checking if the proxy for login_ui is up.

Reply:
But in https://github.com/canonical/identity-platform-login-ui-operator/pull/16/files#diff-8c4de2aa3c77b8aa7f4437d6030f872a02e2fc0ea8bca7fefdeb3dd9413c848dR168 you have 
```python
if any(not data.get(k := key) for key in RELATION_KEYS):
```
Since `default_url` is in the `RELATION_KEYS` wouldn't that fail if `data["default_url"] == ""`?

Reply:
Empty strings are evaluated as false, meaning this check is not needed, because it is already covered in the previous one.

Reply:
deleted highlighted section